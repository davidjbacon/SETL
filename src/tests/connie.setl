$ Connected components of a graph

g := {{1,2},{3,4}};
conn(g);

g := {{1,2},{2,3},{3,4}};
conn(g);

g := {{1,2},{2,3},{4,5}};
conn(g);

PROC conn (g);
  VAR s, ss, t;  $ locals
  print ('Graph: ',g);
  s := {};
  (FOR e IN g)
    t := {ss IN s | EXISTS f IN ss | e*f /= {}};
    s -:= t;
    s +:= {{e}+/t};
  END;
  print ('Connected components: ',s);
END PROC conn;
