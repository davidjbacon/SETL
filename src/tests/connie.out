Graph:  {{1 2} {3 4}}
Connected components:  {{{1 2}} {{3 4}}}
Graph:  {{1 2} {2 3} {3 4}}
Connected components:  {{{1 2} {2 3} {3 4}}}
Graph:  {{1 2} {2 3} {4 5}}
Connected components:  {{{4 5}} {{1 2} {2 3}}}
